group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ResolvedRegularExpression.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.modern.consumer;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * The metadata associated with a regular expression in a Consumer Group.
+ */
+public class ResolvedRegularExpression {
```

Comment:
to understand the usage of this class and the new record:

ConsumerGroupRegularExpressionValue is the log representation of ResolvedRegularExpression. It mainly stores the list of topics that have been resolved with the regex pattern with the latest topic metadata

We store a mapping of regex pattern -> ResolvedRegularExpression (set of topics) which will be used as a cache

Subsequent PRs will
 * try-match existing regex patterns to topics when topic metadata changes
 * create & store new record in log if resolved topics have changed

let me know this is correct

Reply:
Right. Resolving regular expressions is expensive so we decided to store it in the log.

Regarding the subsequent PRs, there are mainly two main areas: 1) We need to resolve the regexes (and refresh them regularly). This process will generate ConsumerGroupRegularExpression records and trigger a rebalance; 2) We need to integrate ResolvedRegularExpression into the target assignment computation.